samples-check/src/main/java/org/gradle/exemplar/test/engine/SamplesRunnerJUnitEngine.java

```diff
@@ -0,0 +1,313 @@
+package org.gradle.exemplar.test.engine;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.gradle.exemplar.executor.CliCommandExecutor;
+import org.gradle.exemplar.executor.CommandExecutionResult;
+import org.gradle.exemplar.executor.CommandExecutor;
+import org.gradle.exemplar.executor.ExecutionMetadata;
+import org.gradle.exemplar.loader.SamplesDiscovery;
+import org.gradle.exemplar.model.Command;
+import org.gradle.exemplar.model.Sample;
+import org.gradle.exemplar.test.normalizer.OutputNormalizer;
+import org.gradle.exemplar.test.SampleModifier;
+import org.gradle.exemplar.test.Samples;
+import org.gradle.exemplar.test.verifier.AnyOrderLineSegmentedOutputVerifier;
+import org.gradle.exemplar.test.verifier.StrictOrderLineSegmentedOutputVerifier;
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.EngineExecutionListener;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.discovery.ClassSelector;
+import org.junit.platform.engine.support.descriptor.AbstractTestDescriptor;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.EngineDescriptor;
+import org.opentest4j.AssertionFailedError;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Collections.emptyList;
+
+public class SamplesRunnerJUnitEngine implements TestEngine {
+
+    // See https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html
+    public static final List<String> SAFE_SYSTEM_PROPERTIES = Arrays.asList("file.separator", "java.home", "java.vendor", "java.version", "line.separator", "os.arch", "os.name", "os.version", "path.separator", "user.dir", "user.home", "user.name");
+
+    private final File tempDir;
+
+    public SamplesRunnerJUnitEngine() throws IOException {
+        tempDir = Files.createTempDirectory("samples-junit-engine").toFile();
+    }
+
+    @Override
+    public String getId() {
+        return "SamplesRunnerJUnitEngine";
+    }
+
+    @Override
+    public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
```

Comment:
Discovery should also support `UniqueIdSelectors` and use `EngineDiscoveryRequestResolver` for de-duplication.